### ISSUE TITLE
Search-Identifier, List and Map's Allowed Types

### ISSUE THREAD

**Issue by hanstzou** [SigmaHQ/sigma-specification#71]
Search-identifier's spec defined value types allowed in lists and maps, but it may need to be expanded.

# Sigma_specification.md

By current definition, **1.** a search-identifier can hold (1) a list of strings, (2) a list of maps, or (3) a map.
**2.** The map, in turn, hold key-value pairs, where each key is a field and each value is (1) a string, (2) an integer, or (3) a null (_special field value_).

> https://github.com/SigmaHQ/sigma-specification/blob/066938f9f62456b21bc8fb77c27d5a6ed5beb175/Sigma_specification.md?plain=1#L421-L424
>
> https://github.com/SigmaHQ/sigma-specification/blob/066938f9f62456b21bc8fb77c27d5a6ed5beb175/Sigma_specification.md?plain=1#L463-L469
>
> https://github.com/SigmaHQ/sigma-specification/blob/066938f9f62456b21bc8fb77c27d5a6ed5beb175/Sigma_specification.md?plain=1#L490-L493
>
> https://github.com/SigmaHQ/sigma-specification/blob/066938f9f62456b21bc8fb77c27d5a6ed5beb175/Sigma_specification.md?plain=1#L555-L560

In addition, although not mentioned directly, there are examples showing that a map's value list also can store integers. Thus, I'd guess a map's value could also be **2.**-(4) a list of strings, integers, or nulls.

> https://github.com/SigmaHQ/sigma-specification/blob/066938f9f62456b21bc8fb77c27d5a6ed5beb175/Sigma_specification.md?plain=1#L500-L505

# Real-World Usages

The aforementioned summarization/guessing is extracted from specification alone.
I checked some Sigma rules in [SigmaHQ/sigma](/SigmaHQ/sigma) repo as well, and found another usage.

In [rules/web/webserver_generic/web_cve_2021_26858_iis_rce.yml#L23-L30](https://github.com/SigmaHQ/sigma/blob/3b11cafc5779db956fc4350a3ece19f53f109847/rules/web/webserver_generic/web_cve_2021_26858_iis_rce.yml#L23-L30) a search-id holds a list which has an integer in it, making an usage of **1.**-(4) a list of strings and integers.
> ``` yml
>     keywords:
>         - 'POST'
>         - 200
>         - '/ecp/DDI/DDIService.svc/SetObject'
> ```

# Are There More?

Will there be other types, say, booleans/floats/...?
For example, is usage of a map like
``` yaml
detection:
    sample-search-id:
        some-field: false
```
considered valid?

# Summary
In conclusion, the spec's type definition seems to be incomplete/inaccurate and may need some update. 

Maybe we can consider defining types as follows
1. A scalar: string/null/boolean/integer/floating point (as defined in [YAML spec](https://yaml.org/spec/1.2.2/#10113-generic-string)).
2. A search-identifier can hold (1) a list of scalars, (2) a list of maps, or (3) a map
3. A map is a key-value pair where the value can be (1) a scalar or (2) a list of scalars.

**Comment by thomaspatzke** [comment]
Agreed. Since pySigma boolean values are also explicitly allowed, therefore it makes sense.
